independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,178 @@
+     */
+    private boolean acquire() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count == 0) {
+                return false;
+            }
+            if (referenceCounter.compareAndSet(count, count + 1)) {
+                return true;
+            }
+        }
+    }
+
+    /**
+     * Decrease the readers counter of the jarFile.
+     * If the counter drops to 0 and a release has been requested also closes the jarFile.
+     *
+     * @return true if the release also closes the underlying jarFile.
+     */
+    private boolean release(JarResource jarResource) {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count <= 0) {
+                throw new IllegalStateException(
+                        "The reference counter cannot be negative, found: " + (referenceCounter.get() - 1));
+            }
+            if (referenceCounter.compareAndSet(count, count - 1)) {
+                if (count == 1) {
+                    try {
+                        jarFile.close();
+                    } catch (IOException e) {
+                        // ignore
+                    } finally {
+                        jarResource.jarFileReference.set(null);
+                    }
+                    return true;
+                }
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Ask to close this reference.
+     * If there are no readers currently accessing the jarFile also close it, otherwise defer the closing when the last reader
+     * will leave.
+     */
+    void close(JarResource jarResource) {
+        release(jarResource);
+    }
+
+    @FunctionalInterface
+    interface JarFileConsumer<T> {
+        T apply(JarFile jarFile, Path jarPath, String resource);
+    }
+
+    static <T> T withJarFile(JarResource jarResource, String resource, JarFileConsumer<T> fileConsumer) {
+
+        // Happy path: the jar reference already exists and it's ready to be used
+        final var localJarFileRefFuture = jarResource.jarFileReference.get();
+        if (localJarFileRefFuture != null) {
+            if (localJarFileRefFuture.join().acquire()) {
+                return consumeSharedJarFile(localJarFileRefFuture, jarResource, resource, fileConsumer);
+            }
+        }
+
+        // There's no valid jar reference, so load a new one
+
+        // Platform threads can load the jarfile asynchronously and eventually blocking till not ready
+        // to avoid loading the same jarfile multiple times in parallel
+        if (!isVirtualThread()) {
+            return consumeSharedJarFile(asyncLoadAcquiredJarFile(jarResource), jarResource, resource, fileConsumer);
+        }
+
+        // Virtual threads needs to load the jarfile synchronously to avoid blocking. This means that eventually
+        // multiple threads could load the same jarfile in parallel and this duplication has to be reconciled
```

Comment:
This means that we are not busy spinning till the load complete, for virtual thread, but waste some RSS (in case of parallel class load over the same jar): I hope we're all in sync that's the choice @Sanne @geoand @dmlloyd

Reply:
I think it's OK for now; hopefully this is all temporary.

Reply:
👌